### PR TITLE
[ch78][ch220][ch221][ch270] API Authorization, Session Management, Connect to Database

### PR DESCRIPTION
**Pull request by LilBank**
https://app.clubhouse.io/cobda/story/220/authentication
https://app.clubhouse.io/cobda/story/270/session-management
https://app.clubhouse.io/cobda/story/221/connect-to-database
https://app.clubhouse.io/cobda/story/78/connect-to-vision-api

## Why?

The quick and complete backend for now. Able to connect to the database and Vision API.

## Changes

- Connected to the database (sqlite3)
- Added user model
- Added product model
- Added authentication methods
- Added session management (next-auth)
- Added image validator (Vision API)

## Screenshot(s)

See [here](https://github.com/Cobda/cobda-web/pull/57/files).

### CODE REVIEW

File: src/components/SignInForm/Form.tsx [hunk 1]
```diff
@@ -1,12 +1,35 @@
...
   const handleSubmitClick = (event: React.MouseEvent<HTMLButtonElement>) => {
     event.preventDefault()
-    router.push('/')
+    signIn('credentials',
+    {
+      email: email,
+      password: password,
+      callbackUrl: '/'
+    })
+  }
```
Comment (kimvcp) [inline src/components/SignInForm/Form.tsx]: Should we use Prettier formatter on this?

```suggestion
    signIn('credentials', {
      email: email,
      password: password,
      callbackUrl: '/'
    })
  }
```

File: src/pages/api/auth/[...nextauth].ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+      credentials: {},
+      authorize: async (credentials) => {
+        const user = await prismaClient.user.findMany({
+          where: {
+            email: credentials.email,
+            password: credentials.password
+          }
+        }).catch(err => err)
+        // TODO: Send useful error message
+        return !user.code && user.length ? user[0] : new Error('The email or password is incorrect.')
```
Comment (kimvcp) [inline src/pages/api/auth/[...nextauth].ts]: Do you think this variable will make your code more legible for other members?

```suggestion
        const isUserAuthorized = !user.code && user.length

        return isUserAuthorized ? user[0] : new Error('The email or password is incorrect.')
```

File: src/pages/api/auth/[...nextauth].ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
+import { NextApiHandler } from 'next'
+import NextAuth from 'next-auth'
+import Providers from 'next-auth/providers'
+import prismaClient from '../../../lib/prisma'
+
+const providers = {
+  providers : [
```
Comment (kimvcp) [inline src/pages/api/auth/[...nextauth].ts]:
```suggestion
  providers: [
```

File: src/prisma/schema.prisma [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+datasource db {
+  provider = "sqlite"
+  url = env("DATABASE_URL")
+}
+
+model User {
+  id    Int     @id @default(autoincrement())
+  email String  @unique
+  password String
+  firstName  String?
```
Comment (kimvcp) [inline src/prisma/schema.prisma]:
```suggestion
  id Int @id @default(autoincrement())
  email String @unique
  password String
  firstName String?
```

**Review by kimvcp — CHANGES_REQUESTED** [review]

File: src/prisma/schema.prisma (hunk 4 above)
Comment (LilBank) [reply src/prisma/schema.prisma]: Good catch!

**Review by kimvcp — APPROVED** [review]

**Review by Kittinske15 — APPROVED** [review]

File: src/prisma/schema.prisma [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+model Product {
+  id Int @id @default(autoincrement())
+  name String
+  size Float
+  description String?
+  price Int
+  owner User? @relation(fields: [ownerId], references: [id])
+  ownerId Int?
```
Comment (Kittinske15) [inline src/prisma/schema.prisma:26]: Follow the ``product-registration``, It should contains color, category and delivery option
```suggestion
  id Int @id @default(autoincrement())
  name String
  color String
  size Float
  description String?
  price Int
  owner User? @relation(fields: [ownerId], references: [id])
  ownerId Int?
  category String
  deliveryOption String
```

**Pull request merged**